lib/pbench/server/api/resources/compare_dataset.py

```diff
@@ -0,0 +1,140 @@
+from pbench.server.api.resources import (
+    APIAbort,
+    ApiAuthorization,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    APIInternalError,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+    Parameter,
+    ParamType,
+    Schema,
+)
+import pbench.server.auth.auth as Auth
+from pbench.server.cache_manager import (
+    CacheManager,
+    TarballNotFound,
+    TarballUnpackError,
+)
+from pbench.server.database.models.datasets import Metadata
+from pbench.server.database.models.users import User
+
+
+class CompareDataset(ApiBase):
+    """
+    This class implements the Server API used to retrieve comparison data for visualization.
+    """
+
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.GET,
+                OperationCode.READ,
+                query_schema=Schema(
+                    Parameter(
+                        "datasets",
+                        ParamType.LIST,
+                        element_type=ParamType.DATASET,
+                        string_list=",",
+                        required=True,
+                    ),
+                ),
+                authorization=ApiAuthorizationType.NONE,
+            ),
+        )
+
+    def _get(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        This function is using Quisby to compare results into a form that supports visualization
+
+        Args:
+            params: includes the uri parameters, which provide the list of dataset.
+            request: Original incoming Request object
+            context: API context dictionary
+
+        Raises:
+            UnauthorizedAccess : The user isn't authorized for the requested access.
+            APIAbort, reporting "NOT_FOUND" and "INTERNAL_SERVER_ERROR"
+            APIInternalError, reporting the failure message
+
+        GET /api/v1/compare?datasets=d1,d2,d3
+        """
+
+        datasets = params.query.get("datasets")
+        authorized_user: User = Auth.token_auth.current_user()
+
+        if not authorized_user:
+            raise APIAbort(
+                HTTPStatus.UNAUTHORIZED,
+                "User provided access_token is invalid or expired",
+            )
+
+        cache_m = CacheManager(self.config, current_app.logger)
+        benchmark = None
+        benchmark_check = set()
+        stream_file = {}
+
+        for dataset in datasets:
+            benchmark = Metadata.getvalue(dataset, "dataset.metalog.pbench.script")
+            benchmark_check.add(benchmark)
+
+            # Validate if all the selected datasets is of same benchmark
+            if len(benchmark_check) != 1:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST,
+                    f"Requested datasets are not of same benchmark. It can't be compared.Benchmarks :{benchmark_check} are requested.",
+                )
```

Comment:
Alternately,
```python
        benchmark_choice = Metadata.getvalue(datasets[0], "dataset.metalog.pbench.script")
        for dataset in datasets:
            benchmark = Metadata.getvalue(dataset, "dataset.metalog.pbench.script")

            # Validate if all the selected datasets is of same benchmark
            if benchmark != benchmark_choice:
                raise APIAbort(
                    HTTPStatus.BAD_REQUEST,
                    f"Requested datasets must all use the same benchmark; found references to {benchmark_choice} and {benchmark}.",
                )
```

Reply:
True enough: I'd had the thought that we could report the disparate set of benchmarks the user had selected. If there are only two, your version is more or less equivalent, although the code is a bit more convoluted. I'm not entirely sure the extra level of diagnostics are important: but if we're going to show more than one benchmark name I'd prefer to show them all.

Reply:
> if we're going to show more than one benchmark name I'd prefer to show them all.

I'm not averse to doing that, but, as you say, we would have to move the check to the end of the loop to make that work (and, if that's what we want, then a `set` is a very good approach).  I was just taking the structure that Siddardh had laid out and making it work with a simple variable instead of a `set`...because it could.  🙂